### PR TITLE
add lint rule for empty enums

### PR DESCRIPTION
**Pull request by DmitryRomaniuk**
https://github.com/denoland/deno_lint/issues/745

### CODE REVIEW

File: src/rules/no_empty_enum.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  }
+}
+
+impl<'c, 'view> Visit for NoEmptyEnumVisitor<'c, 'view> {
+  fn visit_ts_enum_decl(
+    &mut self,
+    enum_decl: &TsEnumDecl,
+    _parent: &dyn Node,
+  ) {
+    if enum_decl.members.len() < 1 {
```
Comment (bartlomieju) [inline src/rules/no_empty_enum.rs:62]:
```suggestion
    if enum_decl.members.is_empty() {
```

File: docs/rules/no_empty_enum.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+Disallows the declaration of an empty enum
+
+An enum with no members serves no purpose. This rule will capture
+these situations as either unnecessary code or a mistaken empty implementation.
+
+### Invalid:
+
+```typescript
+enum Foo {}
+const enum Foo {}
```
Comment (bartlomieju) [inline docs/rules/no_empty_enum.md:10]: Deno doesn't support `const enum`, I suggest to remove this from example as it might be confusing

**Review by bartlomieju — COMMENTED** [review]
Thanks Dmitry! Looking good.

File: docs/rules/no_empty_enum.md (hunk 2 above)
Comment (DmitryRomaniuk) [reply docs/rules/no_empty_enum.md]: thx, fixed

File: src/rules/no_empty_enum.rs (hunk 1 above)
Comment (DmitryRomaniuk) [reply src/rules/no_empty_enum.rs]: thx, fixed

File: src/rules/no_empty_enum.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    if enum_decl.members.is_empty() {
+      self.context.add_diagnostic_with_hint(
+        enum_decl.span,
+        CODE,
+        MESSAGE,
+        HINT,
+      );
+    }
+  }
+}
```
Comment (magurotuna) [inline src/rules/no_empty_enum.rs:71]: If we use `Visit` trait, we have to explicitly call a method to traverse AST nodes further. This means that the current code won't capture the error in the following example:

```ts
enum Foo {
  One = 1,
  Two = (() => {
    enum Bar {} // this won't be detected!
    return 42;
  })(),
}
```

Although this is a valid TypeScript, of course people might not write such a tricky code, but it should be better for the linter to detect it as well.
To address this issue, I recommend to use `VisitAll` trait instead. With this trait, we won't need to do something explicitly, the visitor will traverse all the nodes automatically.
(Or you can also use `Handler` trait - if you're interested in it, see the concrete example, like https://github.com/denoland/deno_lint/blob/9669c247be0e6ce4fcd376acdfd389251e83ed8a/src/rules/no_with.rs)

**Review by magurotuna — COMMENTED** [review]
Nice work!

File: src/rules/no_empty_enum.rs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+use crate::ProgramRef;
+use deno_ast::swc::ast::TsEnumDecl;
+use deno_ast::swc::visit::Node;
+use deno_ast::swc::visit::Visit;
+
+#[derive(Debug)]
+pub struct NoEmptyEnum;
+
+const CODE: &str = "no-empty-enum";
+const MESSAGE: &str = "An empty enum is equivalent to `{{}}`.";
```
[the remark this answers is not in the document]
Comment (DmitryRomaniuk) [reply src/rules/no_empty_enum.rs:12]: thx, fixed

File: src/rules/no_empty_enum.rs (hunk 3 above)
Comment (DmitryRomaniuk) [reply src/rules/no_empty_enum.rs]: I can't use ```VisitAll```, it doesn't work, so I change to ```Handler``` but added if statement, could you review again plz @magurotuna and added that example to test (I've never seen that tricky code before and can't imagine)

File: src/rules/no_empty_enum.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DmitryRomaniuk) [reply src/rules/no_empty_enum.rs]: thx

File: src/rules/no_empty_enum.rs (hunk 3 above)
Comment (DmitryRomaniuk) [reply src/rules/no_empty_enum.rs]: And ci failed on format step 😥

Comment (magurotuna) [reply src/rules/no_empty_enum.rs]: Now the implementation with `Handler` looks perfect 👍 

> And ci failed on format step 😥

We have a script for formatting the code. So just running `./tools/format.ts` will help you.

**Review by magurotuna — APPROVED** [review]
LGTM - thank you!

**Review by bartlomieju — COMMENTED** [review]
LGTM, thanks Dmitry!

**Pull request merged**